### PR TITLE
Created cart PO. Added generic type to return root page after closing. Modified PO creation in other classes

### PR DESCRIPTION
**Pull request by Percifalll**

### CODE REVIEW

File: src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java [hunk 1]
```diff
@@ -1,30 +1,70 @@
...
+    public List<String> getProductNames() {
+        return getProductList().stream()
+                .map(product -> product.$x(".//a[@class='cart-product__title']").text())
+                .collect(Collectors.toList());
+    }
+
+    public BasketModal setProductCountByName(String productName, int count) {
+        SelenideElement countField = getProductByName(productName).$x(".//input[contains(@class, 'cart-counter__input')]");
+        countField.clear();
+        countField.sendKeys(count + "");
```
[the remark this answers is not in the document]
Comment (toleksyn) [reply src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java:49]: String.valueOf()

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java [hunk 2]
```diff
@@ -1,37 +1,78 @@
...
+                .texts();
     }
 
-    public boolean isBasketEmpty() {
-        return getProductsList().isEmpty();
+    public BasketModal setProductCount(String productName, int count) {
+        int totalPrice = getProductsTotalPrice();
+
+        String countFieldXpath = String.format(PRODUCT_XPATH_PATTERN_FOR_NAME, productName) +
+                "//input[contains(@class, 'cart-counter__input')]";
```
Comment (toleksyn) [inline src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java:43]: String.format(PRODUCT_XPATH_PATTERN_FOR_NAME + "//input[contains(@class, 'cart-counter__input')]", productName)
same for others

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java [hunk 3]
```diff
@@ -1,37 +1,79 @@
...
+        $x(productActionXpath).click();
+        $x("//rz-trash-icon/button").click();
+
+        waitForTotalPriceToUpdate(totalPrice);
+        return this;
     }
 
-    private ElementsCollection getProductsList() {
-        return $$x("//li[contains(@class, 'cart-list__item')]");
+    public void close() {
```
Comment (toleksyn) [inline src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java:76]: shout return a PO

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java [hunk 4]
```diff
@@ -1,42 +1,94 @@
...
+    // This pattern should only be used in String.format()
+    // with product name as a second argument
+    private static final String PRODUCT_XPATH_PATTERN_FOR_NAME = "//single-modal-window//li[contains(., '%s')]";
+    private final T rootPage;
 
-    @Step("BasketModal: continue buying products")
-    public ProductPage continueBuying() {
-        $x("//a[contains(@class,'button button_size_medium ')]").click();
-        return new ProductPage();
+    public BasketModal(T rootPage) {
```
Comment (toleksyn) [inline src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java:20]: where is T used?

File: src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java [hunk 5]
```diff
@@ -1,42 +1,94 @@
...
 
-    @Step("BasketModal: get total products price")
-    public int getTotalProductsPrice() {
-        return Integer.parseInt(getTotalProductsPriceWithCurrency().replaceAll("\\D", ""));
+    @Step("BasketModal: check if basket is empty")
+    public boolean isEmpty() {
+        return $(".cart-dummy").is(exist);
+    }
+
+    @Step("BasketModal: get product names ")
```
Comment (toleksyn) [inline src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java:42]: remove space

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java (hunk 4 above)
Comment (toleksyn) [reply src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java:20]: rename

File: src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java [hunk 6]
```diff
@@ -1,42 +1,94 @@
...
+    public List<String> getProductNames() {
```
Comment (toleksyn) [inline src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java:43]: rename

File: src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java [hunk 7]
```diff
@@ -1,42 +1,94 @@
...
+        }
+
+        countField.clear();
+        countField.sendKeys(String.valueOf(count));
+
+        waitForTotalPriceToUpdate(totalPrice);
+        return this;
+    }
+
+    private void waitForTotalPriceToUpdate(int totalPriceBefore) {
```
Comment (toleksyn) [inline src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java:68]: step

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java [hunk 8]
```diff
@@ -1,42 +1,94 @@
...
-    public ProductPage continueBuying() {
-        $x("//a[contains(@class,'button button_size_medium ')]").click();
-        return new ProductPage();
+    public BasketModal(T pageObject) {
+        this.pageObject = pageObject;
     }
 
-    @Step("BasketModal: order products you choose")
-    public OrderingPage orderProducts() {
+    @Step("BasketModal: order products you chose")
```
Comment (vladdmytriv) [inline src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java:24]: selected

Comment (Percifalll) [reply src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java:24]: you was not assigned to review

File: src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java [hunk 9]
```diff
@@ -1,42 +1,94 @@
...
+
+import static com.codeborne.selenide.Condition.exist;
+import static com.codeborne.selenide.Condition.text;
 import static com.codeborne.selenide.Selenide.*;
 
-public class BasketModal {
+public class BasketModal<T> {
+    // This pattern should only be used in String.format()
+    // with product name as a second argument
+    private static final String PRODUCT_XPATH_PATTERN_FOR_NAME = "//single-modal-window//li[contains(., '%s')]";
```
Comment (toleksyn) [inline src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java:17]: if %s in name - template

**Review by toleksyn — CHANGES_REQUESTED** [review]

**Review by NazarYakh — APPROVED** [review]

**Review by NazarYakh — APPROVED** [review]

**Review by dzhyv — APPROVED** [review]

File: src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java [hunk 10]
```diff
@@ -1,42 +1,93 @@
...
-    @Step("BasketModal: get total products price with currency")
-    public String getTotalProductsPriceWithCurrency() {
-        return $x("//div[contains(@class,'sum-price')]").getAttribute("textContent");
+    @Step("BasketModal: get products total price")
+    public int getProductsTotalPrice() {
+        String sum = $x("//div[contains(@class,'sum-price')]//span[1]").text();
+        return Integer.parseInt(sum);
+    }
+
+    @Step("BasketModal: check if basket is empty")
```
Comment (toleksyn) [inline src/main/java/com/ss/ita/rozetka/ui/Modals/BasketModal.java:36]: there's no check here

**Review by toleksyn — CHANGES_REQUESTED** [review]

**Review by vladdmytriv — APPROVED** [review]

**Review by NazarYakh — APPROVED** [review]

**Pull request merged**